File: src/deos/telemetry.py
"""Closed workflow-observation contract shared by platform adapters."""

from __future__ import annotations

from collections.abc import Callable
from datetime import UTC, datetime
from typing import TypeAlias

ObservationValue: TypeAlias = str | int
Observation: TypeAlias = dict[str, ObservationValue]

SCHEMA_VERSION = "1"
STAGES = frozenset(
    {
        "ingress.delivery_record",
        "queue.publish",
        "queue.consume",
        "workflow.transition",
        "linear.issue_update",
    }
)
OUTCOMES = frozenset({"started", "succeeded", "failed", "duplicate"})
ERROR_TYPES = frozenset(
    {
        "correlation_mismatch",
        "d1_operation_failed",
        "queue_publish_failed",
        "linear_transport_failed",
        "linear_http_failed",
        "linear_graphql_failed",
        "unexpected_failure",
    }
)


def workflow_identity(project_id: str, issue_id: str) -> str:
    """Return the canonical run and correlation identifier."""
    if not project_id or not issue_id:
        raise ValueError("project and issue identifiers are required")
    return f"workflow:{project_id}:{issue_id}"


def build_observation(
    *,
    service_name: str,
    stage: str,
    outcome: str,
    correlation_id: str,
    delivery_id: str,
    issue_id: str,
    project_id: str,
    run_id: str,
    error_type: str | None = None,
    message_id: str | None = None,
    attempt_number: int | None = None,
    previous_state: str | None = None,
    next_state: str | None = None,
    cause: str | None = None,
    now: Callable[[], datetime] = lambda: datetime.now(UTC),
) -> Observation:
    """Construct an indexed observation from service-owned, allowlisted values."""
    required_values = {
        "service_name": service_name,
        "correlation_id": correlation_id,
        "delivery_id": delivery_id,
        "issue_id": issue_id,
        "project_id": project_id,
        "run_id": run_id,
    }
    if missing := [name for name, value in required_values.items() if not value]:
        raise ValueError(f"required observation values are empty: {', '.join(missing)}")
    if stage not in STAGES:
        raise ValueError(f"unsupported workflow stage: {stage}")
    if outcome not in OUTCOMES:
        raise ValueError(f"unsupported workflow outcome: {outcome}")
    if outcome == "failed" and error_type not in ERROR_TYPES:
        raise ValueError("failed observations require a supported error type")
    if outcome != "failed" and error_type is not None:
        raise ValueError("error type is allowed only for failed observations")
    if (message_id is None) != (attempt_number is None):
        raise ValueError("Queue message id and attempt number must be supplied together")
    if attempt_number is not None and attempt_number < 1:
        raise ValueError("Queue attempt number starts at 1")

    observation: Observation = {
        "event.time": now().astimezone(UTC).isoformat().replace("+00:00", "Z"),
        "event.name": f"deos.workflow.{stage}",
        "service.name": service_name,
        "deos.telemetry.schema_version": SCHEMA_VERSION,
        "deos.workflow.correlation_id": correlation_id,
        "deos.workflow.stage": stage,
        "deos.workflow.outcome": outcome,
        "linear.delivery.id": delivery_id,
        "linear.issue.id": issue_id,
        "linear.project.id": project_id,
        "deos.workflow.run_id": run_id,
    }
    optional_values: dict[str, ObservationValue | None] = {
        "error.type": error_type,
        "messaging.message.id": message_id,
        "deos.workflow.attempt.number": attempt_number,
        "deos.workflow.previous_state": previous_state,
        "deos.workflow.next_state": next_state,
        "deos.workflow.cause": cause,
    }
    observation.update({key: value for key, value in optional_values.items() if value is not None})
    return observation

File: src/telemetry.ts
export const TELEMETRY_SCHEMA_VERSION = "1" as const;

export const WORKFLOW_STAGES = [
  "ingress.delivery_record",
  "queue.publish",
  "queue.consume",
  "workflow.transition",
  "linear.issue_update",
] as const;

export const WORKFLOW_OUTCOMES = ["started", "succeeded", "failed", "duplicate"] as const;

export const ERROR_TYPES = [
  "correlation_mismatch",
  "d1_operation_failed",
  "queue_publish_failed",
  "linear_transport_failed",
  "linear_http_failed",
  "linear_graphql_failed",
  "unexpected_failure",
] as const;

export type WorkflowStage = (typeof WORKFLOW_STAGES)[number];
export type WorkflowOutcome = (typeof WORKFLOW_OUTCOMES)[number];
export type ErrorType = (typeof ERROR_TYPES)[number];

export interface ObservationInput {
  serviceName: string;
  stage: WorkflowStage;
  outcome: WorkflowOutcome;
  correlationId: string;
  deliveryId: string;
  issueId: string;
  projectId: string;
  runId: string;
  errorType?: ErrorType;
  messageId?: string;
  attemptNumber?: number;
  previousState?: string;
  nextState?: string;
  cause?: string;
}

export type WorkflowObservation = Record<string, string | number>;
export type ObservationWriter = (observation: WorkflowObservation) => void;

export const workflowIdentity = (projectId: string, issueId: string): string => {
  if (!projectId || !issueId) throw new Error("project and issue identifiers are required");
  return `workflow:${projectId}:${issueId}`;
};

export const buildObservation = (
  input: ObservationInput,
  now: () => Date = () => new Date(),
): WorkflowObservation => {
  const required = [
    input.serviceName,
    input.correlationId,
    input.deliveryId,
    input.issueId,
    input.projectId,
    input.runId,
  ];
  if (required.some((value) => value.length === 0)) {
    throw new Error("required observation values must not be empty");
  }
  if (input.outcome === "failed" && input.errorType === undefined) {
    throw new Error("failed observations require an error type");
  }
  if (input.outcome !== "failed" && input.errorType !== undefined) {
    throw new Error("error type is allowed only for failed observations");
  }
  if ((input.messageId === undefined) !== (input.attemptNumber === undefined)) {
    throw new Error("Queue message id and attempt number must be supplied together");
  }
  if (input.attemptNumber !== undefined && input.attemptNumber < 1) {
    throw new Error("Queue attempt number starts at 1");
  }

  return Object.fromEntries(
    Object.entries({
      "event.time": now().toISOString(),
      "event.name": `deos.workflow.${input.stage}`,
      "service.name": input.serviceName,
      "deos.telemetry.schema_version": TELEMETRY_SCHEMA_VERSION,
      "deos.workflow.correlation_id": input.correlationId,
      "deos.workflow.stage": input.stage,
      "deos.workflow.outcome": input.outcome,
      "linear.delivery.id": input.deliveryId,
      "linear.issue.id": input.issueId,
      "linear.project.id": input.projectId,
      "deos.workflow.run_id": input.runId,
      "error.type": input.errorType,
      "messaging.message.id": input.messageId,
      "deos.workflow.attempt.number": input.attemptNumber,
      "deos.workflow.previous_state": input.previousState,
      "deos.workflow.next_state": input.nextState,
      "deos.workflow.cause": input.cause,
    }).filter((entry): entry is [string, string | number] => entry[1] !== undefined),
  );
};

export const writeObservation: ObservationWriter = (observation) => {
  if (observation["deos.workflow.outcome"] === "failed") console.error(observation);
  else console.log(observation);
};
